File: binary trees/BoundaryTraversal.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: a TreeNode
     * @return: a list of integer
     */
    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        List<Integer> boundary = new ArrayList<>();
        if (root == null) return boundary;
        boundary.add(root.val);
        collectLeftBoundary(root.left, boundary);
        collectLeaves(root, boundary);
        collectRightBoundary(root.right, boundary);
        return boundary;
    }
    
    private void collectLeftBoundary(TreeNode node, List<Integer> boundary) {
        if (node == null || isLeaf(node)) return;
        boundary.add(node.val);
        collectLeftBoundary(node.left != null ? node.left : node.right, boundary);
    }
    
    private void collectRightBoundary(TreeNode node, List<Integer> boundary) {
        if (node == null || isLeaf(node)) return;
        collectRightBoundary(node.right != null ? node.right : node.left, boundary);
        boundary.add(node.val);
    }
    
    private void collectLeaves(TreeNode node, List<Integer> boundary) {
        if (node == null) return;
        if (isLeaf(node)) {
            boundary.add(node.val);
        } else {
            collectLeaves(node.left, boundary);
            collectLeaves(node.right, boundary);
        }
    }
    
    private boolean isLeaf(TreeNode node) {
        return node.left == null && node.right == null;
    }
}

File: binary trees/BoundaryTraversal.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * @param root: a TreeNode
     * @return: a list of integer
     */
	void printLeaves(TreeNode *root, vector<int> &res){
		if(!root) return;
		if(!root->left && !root->right) res.push_back(root->val);
		printLeaves(root->left, res);
		printLeaves(root->right, res);
	}

	void printLeft(TreeNode *root, std::vector<int> &res) {
		if(!root) return;
		res.push_back(root->val);
		if(!root->left) {
			printLeft(root->right, res);
		}else{
			printLeft(root->left, res);	
			printLeaves(root->right, res);		    
		}
	}

	void printRight(TreeNode *root, std::vector<int> &res) {
		if(!root) return;
		if(!root->right) {
			printRight(root->left, res);
		}else{
			printLeaves(root->left, res);
			printRight(root->right, res);
		}
		
		res.push_back(root->val);
	}

	vector<int> boundaryOfBinaryTree(TreeNode * root) {
		std::vector<int> res;
		if(!root) return res;
		res.push_back(root->val);
		printLeft(root->left, res);
		printRight(root->right, res);
		return res;
	}
};	